Reset connection pool upon failure to read response

ahojnnes opened the pull request:
This PR fixes the issues reported in https://github.com/Azure/azure-sdk-for-cpp/issues/4206 when talking to azurite. It appears that it can happen that all the connections against the same connection key in the pool may be broken/unusable. If there are many such broken connections in the pool (i.e., `> MaxRetries`), the request will eventually fail, because the existing retry logic would attempt to execute the same request against the broken set of connections. Note that the code to reset the pool was never reached, because the exception would exit the loop early where the reset would happen later. The approach in this PR avoids this issue by not throwing an exception but returning a CURLcode for `ReadStatusLineAndHeadersFromRawResponse`. I am not sure if this is the proper approach. I tested this against azurite and real blob storage. It seems to behave correctly.

Open questions:
* The same logic may have to be applied to the `ReadExpected` function. I have not had time to figure out all the places where this is used though and what the appropriate approach would be. Also return a CURLcode?
* There are some (orthogonal) improvements to `ExtractOrCreateCurlConnection` which would early discard connections that are shutdown/expired. These changes are not required to fix the original issue referenced above. However, it seems like that logic could alleviate the need for spawning a separate cleanup thread?

Jinming-Hu commented:
Hi @ahojnnes 

> There are some (orthogonal) improvements to ExtractOrCreateCurlConnection which would early discard connections that are shutdown/expired. These changes are not required to fix the original issue referenced above. However, it seems like that logic could alleviate the need for spawning a separate cleanup thread? I can revert those changes.

Are you referring to the changes at Line 2000+ in `curl.cpp`? IIRC, we ever came across an issue where closing the connection may take very long time (very rare, but it did happen). So, please do not destruct any curl sessions when holding the mutex.

ahojnnes commented:
> Are you referring to the changes at Line 2000+ in curl.cpp? 

Yes, I was referring to those changes. I'll revert those changes then.

Jinming-Hu reviewed (approved):
LGTM

antkmsft commented:
/azp run cpp - core - ci

antkmsft commented:
@LarryOsterman, do changes look good to you?

ahojnnes commented:
The CI is failing but the failures seem unrelated to the code changes in this PR?

antkmsft commented:
@ahojnnes, yes. No CI problems with your changes. We have a broader issue with CI at the moment, some other PRs were also affected.

antkmsft commented:
/azp run cpp - core - ci

LarryOsterman reviewed (approved):
My only concern here is that with this change, we lose a bit of important information about the cause of the failure.

It might make sense to include a Log::Write in the places where a `throw` was turned into `return CURL_E_RECV_ERROR;`

vhvb1989 commented:
@ahojnnes if the connection in the pool is broken, why is `auto result = SendRawHttp(context);` not returning an error?

`SendRawHttp` should be the first to return an error if the connection is not usable any more (like if it was closed already by the server).  I am curious to know why this function is working and the error is not happening until we are trying to read back something from the service.  

To me, this is behavior is not enough to assume that the connection is broken.  If it is really broken by the time we want to re-use it, we should debug what's happening during `sendRawHttp()` and why it is not failing.

I don't think this change is a good idea, because it will hide cases where there connection is still working but there is a real error (like a network disconnection or a service error) while reading the response. In that case, we should rely on the RetryPolicy from the Http pipeline and not on the connection pool.  
We could end up blocking the application forever if there is a bug in the server or the network and we just keep trying to create a new connection (thinking that the previous connection was broken).

If this is a unique case for Azurite, can it work for you to disable `keep-alive` feature ?  
This way, the connection pool won't keep the connections to be re-used and you will always create a new one.
If performance is not important, you could explore this path

Jinming-Hu commented:
@vhvb1989 I **guess** it's because `SendRawHttp` only writes into the socket. The data is not necessarily sent over the wire when the write returns. The data may be buffered in the kernel/NIC etc. But read operation is different. If there's no data in the local buffer, it has to wait until data comes from the wire.

ahojnnes commented:
Thanks for the responses! I addressed @LarryOsterman's comments and added error log statements for those cases. I admittedly don't know enough about the inner workings of the kernel to make a confident statement about @vhvb1989's concerns.

ahojnnes commented:
@vhvb1989 How would I disable "keep-alive" with the SDK?

Jinming-Hu commented:
> How would I disable "keep-alive" with the SDK?

There's an option. But I don't think you want to do that. As I understand it, you're running against Azurite for testing, and running against Azure Storage in production right? Disabling keep-alive will decrease the performance.

ahojnnes commented:
Yes, that's right.

vhvb1989 commented:
@Jinming-Hu , SendRawHttp() calls `CurlConnection::SendBuffer()`, which calls `curl_easy_send()` (see:  https://curl.se/libcurl/c/curl_easy_send.html)

If the state of response from the socker is not enough to find out if the connection is still valid (not closed), we need to find another way to find out if the connection was closed.  As I mentioned before, using the `read` response/result to detect if the connection is closed will bring other side effects (issues on other scenarios).  

@LarryOsterman , do you know if there's any way to check the connection (may be from the socket id)?   Ideally, the best scenario is if we can test it even before calling `SendRawHttp()`

LarryOsterman commented:
> @Jinming-Hu , SendRawHttp() calls `CurlConnection::SendBuffer()`, which calls `curl_easy_send()` (see: https://curl.se/libcurl/c/curl_easy_send.html)
> 
> If the state of response from the socker is not enough to find out if the connection is still valid (not closed), we need to find another way to find out if the connection was closed. As I mentioned before, using the `read` response/result to detect if the connection is closed will bring other side effects (issues on other scenarios).
> 
> @LarryOsterman , do you know if there's any way to check the connection (may be from the socket id)? Ideally, the best scenario is if we can test it even before calling `SendRawHttp()`

This is a tricky problem - the way TCP (and sockets) work is that unless the remote side sends a termination frame to the sender, the client has no way of knowing if the connection is gone. Usually the only way is to send data and see if it succeeds. And even then, the TCP stack will often buffer requests locally before actually sending them on the wire.

vhvb1989 commented:
> @vhvb1989 How would I disable "keep-alive" with the SDK?

@ahojnnes, if you are using some env-var for switching between PROD/TEST/DEV, you can use that env-var in your code to only set `keep-alive=false` for TEST/DEV and not PROD. You just need to use the `clientOptions` for the SDK client that you are creating. For example, if you are creating a blob client, it would look something like:

```cpp
       BlobClientOptions clientOptions;
      
      // For non-prod envs (Azurite or other) do not use the curl pool to re-use connections.
      // Instead, create a new connection for every request.
      // Perform will be lower in the scale of seconds.
      // This is to workaround an issue where Azurite closes connections sooner than the Actual Azure Storage service.
      if (std::getenv("APP_ENV") != "PROD") {
        Azure::Core::Http::CurlTransportOptions curlOptions;
        curlOptions.HttpKeepAlive = false;
        clientOptions.Transport.Transport = std::make_shared<Azure::Core::Http::CurlTransport>(curlOptions);
      }

      auto containerClient = BlobContainerClient::CreateFromConnectionString("connString","container", clientOptions);
```

The clientOptions is an optional parameter for any Sdk client constructor, and it defines all the client options, including the transport adapter. When you don't set anything, the default transport adapter is curl with `keepAlive=true` (on Linux).

> Note:  You will need to include the header `#include "azure/core/http/curl_transport.hpp"` in the example above.

The transport adapter headers are not part of the common header `core.hpp` that is included by services.
If you are using Windows, you need to set `curl` transport adapter to use (otherwise you would get winhttp)

ahojnnes commented:
> This is a tricky problem - the way TCP (and sockets) work is that unless the remote side sends a termination frame to the sender, the client has no way of knowing if the connection is gone. Usually the only way is to send data and see if it succeeds. And even then, the TCP stack will often buffer requests locally before actually sending them on the wire.

This is the issue that this PR is attempting to address (as far as I understand) but I don't know if there is a better way. It appears that the current curl wrapper is not robust to this scenario. It surfaced when running against azurite but the same could happen against real Azure as well? The proposed changes in this PR resolve the issues I see against azurite. I honestly don't think it is a reasonable proposal to disable the "keep-alive" option to work around this issue, as it comes with other trade offs.

Jinming-Hu commented:
Hi @vhvb1989 , I looked through the source code again today. `ReadFromSocket` internally calls `curl_easy_recv()` and returns the number of bytes read from the socket when the return code of `curl_easy_recv()` is `CURL_OK`. In the API reference of [curl_easy_recv](https://curl.se/libcurl/c/curl_easy_recv.html), it says 

```
Reading exactly 0 bytes indicates a closed connection.
```

So I think it's safe to consider the connection is not available anymore when `this->m_innerBufferSize` equals to zero.
I also think it's reasonable to follow the same processing logic as `SendRawHttp` returns error, which also indicates the connection is broken.

ahojnnes commented:
@Jinming-Hu Thanks, the "Reading exactly 0 bytes indicates a closed connection." part lead me to make the proposed changes in this PR. The one open question that I had is whether the other calls to `ReadFromSocket` in curl.cpp, which are not updated in this PR would also be affected. As far as I can tell, they are not affected, as those calls are not wrapped in the connection pool retry loop in `CurlTransport::Send`. I appreciate if somebody can double verify those places.

LarryOsterman commented:
/azp run cpp - core - ci

LarryOsterman commented:
/check-enforcer bypass

LarryOsterman commented:
/check-enforcer override
